move: support automated address management

rvantonder opened the pull request:
## Description 

This PR adds automated address management support for `sui client publish` and `sui client upgrade`. The `chain_id` is resolved early in sui commands and propagated down where published ID resolution is needed.

High-level:

- If a package contains a `Move.lock` that contains info to support a package upgrade, then that is used, and a user does not have to (nor should they) maintain a `published-at` address in their `Move.toml`.

- If a package does _not_ have any `Move.lock` information to support a package upgrade, we will still fallback to `Move.toml` as we do today. Nothing changes.

- If a package has both a `Move.lock` and `Move.toml`, then the `Move.lock` will be used (the user is free to delete `published-at` in their `Move.toml`). Caveat: If a published address is in both of these files, they must be the same (we will raise a `Conflict` error message otherwise).
 
- Note: docs to follow and merge at the same time as this PR, particularly [this section on upgrades](https://docs.sui.io/concepts/sui-move-concepts/packages/upgrade#example)

- Note after this PR it is still required to set `[addresses]` to `0x0` in `Move.toml`. This requirement will be removed subsequently (which also requires setting to `0x0` still). I didn't want to meddle with that logic all at once.

## Test plan 

How did you test the new or updated feature?

We cover three cases:

- Upgrade with `Move.toml` only (existing package upgrade test)
- Upgrade with `Move.lock` only (extension of existing package management test)
- Upgrade with `Move.lock` and `Move.toml` with conflicting addresses error condition (new test)

- Existing e2e tests pass (and SDK tests now resolve chain ID for publish commands)

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [x] CLI: Running `sui client publish` or `sui client upgrade` will now cause package addresses to be tracked in the `Move.lock` file. This removes the need to manually record or edit `published-at` addresses in the `Move.toml` file going forward. See docs PR https://github.com/MystenLabs/sui/pull/18433 for full details.
- [ ] Rust SDK:

rvantonder commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -1960,6 +1930,128 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
+    let mut lines: Vec<String> = buf.split('\n').map(|x| x.to_string()).collect();
+    let idx = lines.iter().position(|s| s == "[package]").unwrap();
+    // Purposely add a conflicting `published-at` address to the Move manifest.
+    lines.insert(idx + 1, "published-at = \"0xbad\"".to_string());
```
Main logic for this test: put a bogus address in `Move.toml` and expect the error message.

rvantonder commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -1872,29 +1863,8 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
-        .unwrap();
     upgrade_pkg_path.pop();
 
-    let mut buf = String::new();
-    move_toml.read_to_string(&mut buf).unwrap();
-
-    // Add a `published-at = "0x<package_object_id>"` to the Move manifest.
-    let mut lines: Vec<String> = buf.split('\n').map(|x| x.to_string()).collect();
-    let idx = lines.iter().position(|s| s == "[package]").unwrap();
-    lines.insert(
-        idx + 1,
-        format!(
-            "published-at = \"{}\"",
-            package.reference.object_id.to_hex_uncompressed()
-        ),
-    );
-    let new = lines.join("\n");
-    move_toml.write_at(new.as_bytes(), 0).unwrap();
-
```
This test `test_package_management_on_upgrade_command` previously performed an upgrade to check that we are writing the correct package management info to `Move.lock`. It previously used the same logic, relying on `Move.toml` to have the info.

Now, we remove the reliance on `Move.toml` to perform the upgrade in this test (and it ought to still succeed).

rvantonder commented on crates/sui-move-build/src/lib.rs:
```diff
@@ -649,16 +653,18 @@ pub enum PublishedAtError {
 
     for (name, package) in &resolution_graph.package_table {
-        let property = published_at_property(package);
+        let property = resolve_published_id(package, chain_id.clone());
```
Main top-level change: previously we resolved the `published_at` property only from the manifest. Now we do so from either the `Move.lock` or manifest in the `resolve_published_id` function.

rvantonder commented on sdk/typescript/test/e2e/utils/setup.ts:
```diff
@@ -68,9 +70,15 @@ export function getClient(url = DEFAULT_FULLNODE_URL): SuiClient {
 	return setupWithFundedAddress(keypair, address, options);
 }
 
+export async function setupClientConfig() {
+	const configPath = path.join(homedir(), '.sui', 'sui_config', 'client.yaml');
+	execSync(`${SUI_BIN} client --yes --client.config ${configPath}`, { encoding: 'utf-8' });
+}
+
```
SDK e2e tests need a `client.config` path to exist in the environment, where the `Move.lock` will be written to.

hayes-mysten replied:
can we use something other than the home directory for this?  Using a users config from their home directory seems like a bad idea of e2e tests (when running locally)

hayes-mysten replied:
eg, Either using a temp directory, or creating something in the repo with .gitignore to avoid checking it in

rvantonder replied:
Ah, I was basing this off of what we currently do in [here](https://sourcegraph.com/github.com/MystenLabs/sui@dd95851ea79ba88b8c9aba8d88aab9e86657ebec/-/blob/examples/trading/api/sui-utils.ts?L27-29) which uses `homedir()`, so it seemed reasonable to use that for client config too?

hayes-mysten replied:
I don't think that ever gets run as part of our e2e test suit, I think that is just an example app someone created, which is a little different

rvantonder replied:
Aha, makes sense, I'll look into it!

rvantonder commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -3021,9 +3113,11 @@ async fn test_get_owned_objects_owned_by_address_and_check_pagination() -> Resul
 
 #[tokio::test]
 async fn test_linter_suppression_stats() -> Result<(), anyhow::Error> {
+    let mut test_cluster = TestClusterBuilder::new().build().await;
+    let context = &mut test_cluster.wallet;
```
Updates this test to use a local network to resolve a `chain_id` (required to resolve addresses from `Move.lock`)

stefan-mysten commented on crates/sui-package-management/src/lib.rs:
```diff
@@ -73,3 +88,70 @@ pub async fn update_lock_file(
+    let published_id_in_manifest = match published_at_property(package) {
+        Ok(v) => Some(v),
+        Err(PublishedAtError::NotPresent) => None,
+        Err(e) => return Err(e), // An existing but invalid `published-at` in `Move.toml` should fail early.
```
nit: what error would this return?

rvantonder replied:
This is a parsing error, if the `published-at` isn't a valid hexadecimal address (same behavior as before, but simply propagated early here).

stefan-mysten commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -134,7 +134,6 @@ async fn test_addresses_command() -> Result<(), anyhow::Error> {
     let test_cluster = TestClusterBuilder::new().build().await;
     let mut context = test_cluster.wallet;
 
-    // Add 3 accounts
```
// Add 4 accounts

rvantonder replied:
Hmm, I think this line got removed by accident, not sure how. It is 3 accounts (loop inclusive of 0, exclusive of 3), so I will just revert the change.

stefan-mysten commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -1849,20 +1848,12 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
         .find(|refe| matches!(refe.owner, Owner::AddressOwner(_)))
         .unwrap();
 
-    // Hacky for now: we need to add the correct `published-at` field to the Move toml file.
```
nit: add a TODO here

stefan-mysten commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -1960,6 +1930,128 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
+        .await?
+        .data;
+
+    // Check log output contains all object ids.
```
nit: remove comment or fix it

stefan-mysten reviewed (approved):
Looking forward to seeing this merged!

rvantonder commented on crates/sui-package-management/src/lib.rs:
```diff
@@ -3,16 +3,31 @@
+use move_symbol_pool::Symbol;
 use sui_json_rpc_types::{get_new_package_obj_from_response, SuiTransactionBlockResponse};
-use sui_sdk::wallet_context::WalletContext;
+use sui_sdk::{types::base_types::ObjectID, wallet_context::WalletContext};
```
Good call! Fixed up to use sui_types

awelc commented on crates/sui/src/sui_commands.rs:
```diff
@@ -339,8 +343,22 @@ impl SuiCommand {
+                    client_config.unwrap_or(sui_config_dir()?.join(SUI_CLIENT_CONFIG));
+                prompt_if_no_config(&config_path, accept_defaults).await?;
+                let context = WalletContext::new(&config_path, None, None)?;
+                let chain_id = context
```
As I understand it, this is querying the network but what happens if there is no network connectivity? Does it failover gracefully and uses `None` for `chain_id` or we are getting problems with off-line execution of Move commands?

awelc reviewed (changes requested):
Overall looks good but keeping off-line execution of Move commands seems important

rvantonder commented:
> Overall looks good but keeping off-line execution of Move commands seems important

Move commands aren't entirely offline though? We need to check for and pull in dependendencies. Caching helps but it doesn't remove the general online requirement.

There may be some rationale to keep some of the build process offline but whatever that looks like I suspect it is not going to be smooth once we introduce `.move` resolvers and compiling on-chain deps. 

I'll pull in more discussion here from slack, but it seems somewhat unavoidable in the long run. A fallback to `None` effectively disables the ability to automatically manage addresses (we will never know which chain's dependencies to use). So it will only be useful in a testing / debug / local setting (for which we need a connection to some network in most contexts already). For example there was only one `sui move` test in the repo that needed to be updated to use a local network for testing.

awelc commented:
The reasoning makes sense. Do we run `sui move build` during `cargo build` for Sui, right? Loosing the ability to build Sui off-line would (selfishly) be rather inconvenient.

rvantonder commented:
@awelc fair, I think there might be some ways to help with that. I'll fold it into the discussion.

awelc commented:
> @awelc fair, I think there might be some ways to help with that. I'll fold it into the discussion.

Thank you! I am sorry to be a bit (?) picky about this, but I am also not going to stand in the way of progress if it's decided that sacrifices need to be made

awelc reviewed (commented):
Removing "request changes" mark after raising some issues so that it's in a land-able state

amnn reviewed (commented):
Copying parts of a thread about `sui move build` and always requiring a network connection, because I think that's where we're at with this PR -- let me know if I'm operating on stale info though!

From me:
> My take is that it's fine to expect to be able to talk to the RPC in question if you are publishing but not if you are just building, and as far as I can tell, the only substantive reason you would need to talk to RPC before the build is to do address assignment, which is only relevant if the results are to be published -- is that a fair assessment?
>
> @rvantonder is [suggesting]:
> 1. On first build, fetch / update dependencies.
> 2. Every time: Query RPC for chain identifier, and use this to fetch address assignments out of the fetched dependencies.
> 3. Complete resolution and send to compiler.
>
> The revision I'm suggesting is:
>
> 1. On first build, fetch / update dependencies.
> 2. If building to publish
>    1. then, query RPC for chain identifier and use this to fetch address assignments out of the fetched dependencies.
>    2. otherwise, make up some addresses for your dependencies, because honestly who cares.
> 3. Complete resolution and send to compiler.

From @rvantonder:
> @amnn your revision sounds good, and [2(ii)] is the important step.
>
> Right now we unconditionally resolve published IDs via Move.toml (both move build and client publish).
>
> Agree that it does not actually matter for the build step, though I didn't think ahead to separate this out and/or use "throwaway" addresses.
>
> I went the other route of trying to unconditionally and legitimately resolve IDs for both build or publish workflows.
>
> Going to give this a try and modify to have the build logic not rely on resolving published IDs at all.

rvantonder commented:
@amnn yes, accurate! Working on fixing up the `sui move build` case. Will re-request when ready!

rvantonder commented:
Please take another look!

@hayes-mysten could you throw an eye on the TypeScript changes? I think I addressed your feedback here (using a `tmpt` folder and other nits).

@awelc @amnn it looks like we don't need to populate any published IDs for the `move build` case after all, that code path is (silently) resilient to the case where published IDs are not resolved (as with `Move.toml` before). I don't love that it's silent and might prefer that we explicitly resolve published IDs only when we need to, but that comes at the expense of more complicated changes. I may revisit in future if we think it's worth separating those concerns.

amnn commented on crates/sui-package-management/src/lib.rs:
```diff
@@ -73,3 +90,65 @@ pub async fn update_lock_file(
+    };
+
+    let lock = package.package_path.join(SourcePackageLayout::Lock.path());
+    let mut lock_file = match File::open(lock.clone()) {
+        Ok(f) => f,
+        Err(_) => match published_id_in_manifest {
+            Some(v) => {
+                return ObjectID::from_str(v.as_str())
+                    .map_err(|_| PublishedAtError::Invalid(v.to_owned()))
+            }
+            None => return Err(PublishedAtError::NotPresent),
+        },
+    };
```
nit (optional): `let ... else`?

```suggestion
    let Ok(mut lock_file) = File::open(lock.clone()) else {
        return match published_id_in_manifest {
            Some(v) => {
                return ObjectID::from_str(v.as_str())
                    .map_err(|_| PublishedAtError::Invalid(v.to_owned()))
            }
            None => return Err(PublishedAtError::NotPresent),
        },
    };
```

rvantonder replied:
Done!

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1631,12 +1652,25 @@ pub(crate) async fn upgrade_package(
+            "Invalid 'published-at' field in Move.toml or Move.lock of package to be upgraded. \
                          Expected an on-chain address, but found: {v:?}"
         ),
+        PublishedAtError::Conflict(id_lock, id_manifest) => {
```
nit: you may want to make `Conflict` have named fields so it becomes more difficult to mistake the meanings of these two values, seeing as they are the same type.

amnn replied:
Really like how clear this error message is, thanks for this!

amnn reviewed (commented):
I'm glad that you were able to avoid the chain identifier call outside of publishing and upgrading @rvantonder! One thing that I'm not clear on though is how this implementation deals with just the regular named address to numerical address mapping (the one that is currently fed from the `[addresses]` section of `Move.toml`).

My understanding was that automated address management would also supplant this, using the original IDs for all packages as their `[address]` (otherwise people will still need to update their `Move.toml` file after publishing, with their package's Original ID), but then I can't spot logic to compensate for a lack of chain identifier to pick an original ID from in the `move build` path.

rvantonder commented:
> My understanding was that automated address management would also supplant this, using the original IDs for all packages as their [address]

Yes indeed! I put a note in the PR description, expect this to be up soon.

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1631,12 +1652,25 @@ pub(crate) async fn upgrade_package(
+            "Invalid 'published-at' field in Move.toml or Move.lock of package to be upgraded. \
                          Expected an on-chain address, but found: {v:?}"
         ),
+        PublishedAtError::Conflict{id_lock, id_manifest} => {
```
nit: `cargo fmt` (I think it will complain at you in CI anyway).

amnn commented on crates/sui-source-validation-service/src/lib.rs:
```diff
@@ -151,6 +151,14 @@ pub async fn verify_package(
     package_path: impl AsRef<Path>,
 ) -> anyhow::Result<(Network, AddressLookup)> {
     move_package::package_hooks::register_package_hooks(Box::new(SuiPackageHooks));
+    let network_url = match network {
+        Network::Mainnet => MAINNET_URL,
+        Network::Testnet => TESTNET_URL,
+        Network::Devnet => DEVNET_URL,
+        Network::Localnet => LOCALNET_URL,
+    };
```
If you're going to address this in the same stack no worries, but otherwise, can you add a TODO for this in the code?

amnn reviewed (approved):
🚀 Exciting times! Looking forward to getting this in people's hands, thanks @rvantonder !
